File: md-ml-py/matmul_fake_offline.py
"""Fake offline phase for matrix multiplication experiment."""

import time
from md_ml.share import Spdz2kShare64
from md_ml.fake_offline import FakeParty, FakeCircuit

FAKE_OFFLINE_DIR = "../fake-offline-data"
JOB_NAME = "MatMul-py"
DIM = 4096


def main():
    share = Spdz2kShare64

    print(f"Running fake offline phase for {DIM}x{DIM} matmul...")
    t = time.time()

    party = FakeParty(share, JOB_NAME, FAKE_OFFLINE_DIR, n_parties=2)
    circuit = FakeCircuit(share)

    a = circuit.input(0, DIM, DIM)
    b = circuit.input(0, DIM, DIM)
    c = circuit.multiply(a, b)
    d = circuit.output(c)

    circuit.add_endpoint(d)
    circuit.run_offline(party)

    print(f"Fake offline phase done in {(time.time()-t)*1000:.0f} ms")


if __name__ == "__main__":
    main()

File: md-ml-rs/src/bin/matmul_fake_offline.rs
// Matrix multiplication experiment - fake offline phase (Rust rewrite)

use md_ml::share::Spdz2kShare64;
use md_ml::fake_offline::{FakeParty, FakeCircuit};

const FAKE_OFFLINE_DIR: &str = "../fake-offline-data";
const JOB_NAME: &str = "MatMul";
const DIM: usize = 4096;

fn main() {
    type S = Spdz2kShare64;

    let mut party = FakeParty::<S, 2>::new(JOB_NAME, FAKE_OFFLINE_DIR);
    let mut circuit = FakeCircuit::<S, 2>::new();

    let a = circuit.input(0, DIM, DIM);
    let b = circuit.input(0, DIM, DIM);
    let c = circuit.multiply(a, b);
    let d = circuit.output(c);

    circuit.add_endpoint(d);
    circuit.run_offline(&mut party);
}
